### PR TITLE
KEYCLOAK-13616 Setup for the product

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.redhat.com/browse/KEYCLOAK-13616

This Pull Request introduces a setup required for building both Keycloak and RHSSO from the same repo. OSBS (which is used by the product) requires additional YAML files (added by this commit). I also introduced `productize.sh` script, which replaces `Dockerfile` with the on prepared specially for OSBS.

### CODE REVIEW

**Comment by slaskawi** [comment]
@ASzc @hmlnarik @mhajas FYI

**Comment by stianst** [comment]
What is an downstream sync PR?! Changes go downstream, not upstream.

**Comment by slaskawi** [comment]
@stianst We are syncing the code with changes in the downstream repo. Therefore, it's a downstream sync ;)

**Comment by stianst** [comment]
Huh? Changes are applied upstream, then synced downstream. Not the other way around.

**Comment by slaskawi** [comment]
@stianst I changes this to `OSBS sync`. Does this work for you?

**Comment by stianst** [comment]
I take it OSBS requires the two extra yaml files to build the Operator? If so I would say a title "Add metadata files needed by OSBS" would be much better.

Lack of information in the title/description and just pointing to a large JIRA makes it really hard and time consuming to review things.

**Comment by slaskawi** [comment]
@stianst Thanks for the comment!

I updated the commit as well as the description for this Pull Request. I also introduced the `productize.sh` script, which works a bit differently than what we agreed. I found putting labels a bit problematic (I had to index the lines), so in this proposal, I introduces a separate `Dockerfile.rhel8` file. I think this way it will be a bit less error prone.

**Comment by slaskawi** [comment]
As per @hmlnarik request, I modified rendering product `Dockerfile` to a script. The result looks like the following:

```
$ cat Dockerfile 
FROM ubi8-minimal:8-released AS build-env

RUN microdnf install -y git make golang

COPY . /src/
RUN cd /src && make code/compile
RUN cd /src && echo "Build SHA1: $(git rev-parse HEAD)"
RUN cd /src && echo "$(git rev-parse HEAD)" > /src/BUILD_INFO

# final stage
FROM ubi8-minimal:8-released

LABEL com.redhat.component="redhat-sso-7-sso74-operator-rhel8-container"  description="Red Hat Single Sign-On 7.4 Operator on OpenJDK OpenShift container image, based on the Red Hat Universal Base Image 8 Minimal container image" summary="Red Hat Single Sign-On 7.4 Operator on OpenJDK OpenShift container image, based on the Red Hat Universal Base Image 8 Minimal container image" version="7.4" io.k8s.description="Operator for Red Hat SSO" io.k8s.display-name="Red Hat SSO 7.4 Operator" io.openshift.tags="sso,sso74,keycloak,operator" name="rh-sso-7/sso74-operator-rhel8" maintainer="Red Hat Single Sign-On Team"

COPY --from=build-env /src/BUILD_INFO /src/BUILD_INFO
COPY --from=build-env /src/tmp/_output/bin/keycloak-operator /usr/local/bin

ENTRYPOINT ["/usr/local/bin/keycloak-operator"]
```

File: Dockerfile [hunk 1]
```diff
@@ -2,15 +2,15 @@ FROM registry.access.redhat.com/ubi8/ubi-minimal:8.1 AS build-env
...
-ADD ./ /src
+COPY . /src/
 RUN cd /src && make code/compile
 RUN cd /src && echo "Build SHA1: $(git rev-parse HEAD)"
 RUN cd /src && echo "$(git rev-parse HEAD)" > /src/BUILD_INFO
 
 # final stage
 FROM registry.access.redhat.com/ubi8/ubi-minimal:8.1
 
-RUN microdnf update && microdnf clean all && rm -rf /var/cache/yum/*
```
Comment (hmlnarik) [inline Dockerfile:13]: Why is this step removed?

File: productize.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash -ex
+
+LABELS=$(cat <<-END
+LABEL \
+com.redhat.component="redhat-sso-7-sso74-operator-rhel8-container"  \
+description="Red Hat Single Sign-On 7.4 Operator on OpenJDK OpenShift container image, based on the Red Hat Universal Base Image 8 Minimal container image" \
```
Comment (hmlnarik) [inline productize.sh:6]: Please remove reference to JDK since this is not relevant.

File: productize.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+io.k8s.description="Operator for Red Hat SSO" \
+io.k8s.display-name="Red Hat SSO 7.4 Operator" \
+io.openshift.tags="sso,sso74,keycloak,operator" \
+name="rh-sso-7\/sso74-operator-rhel8" \
+maintainer="Red Hat Single Sign-On Team"
+END
+)
+
+sed -i 's/registry.access.redhat.com\/ubi8\/ubi-minimal:8.1/ubi8-minimal:8-released/' Dockerfile
+sed -i "s/##LABELS/$LABELS/g" Dockerfile
```
Comment (hmlnarik) [inline productize.sh:18]:
```suggestion
sed -i -e 's/FROM registry.access.redhat.com\/ubi8\/ubi-minimal:[0-9.]*/FROM ubi8-minimal:8-released/' \
    -e "s/##LABELS/$LABELS/g" \
    Dockerfile
```

File: container.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+compose:
+  pulp_repos: true
+go:
+  modules:
+    - module: github.com/keycloak/keycloak-operator
+image_build_method: imagebuilder
+platforms:
+  only:
+    - x86_64
+    - ppc64le
```
Comment (hmlnarik) [inline container.yaml:10]: Comment out ppc

File: Dockerfile (hunk 1 above)
Comment (slaskawi) [reply Dockerfile:13]: I realized this doesn't buy us anything.

In case of a CVE in UBI8, we get an updated package (with a CVE fix) via UBI8 image respin. Then, we need to respin our image to pick the fix up. We also don't rely on any upgraded package, so performing `microdnf update` doesn't help us in any sense. 

However, this has at least two pretty big drawbacks. At first it creates a new layer of RPMs, which may (depending on the amount of upgraded packages) increase the image size (moreover, the image size becomes unpredictable). It is far more beneficial to get those upgrades via a respun UBI8 image. The `microdnf update` procedure also has some (I guess very minimal, but still) impact on image integrity. We can be sure that pure UBI8 image has been verified by RHEL QE, but I wouldn't be so sure if they do verify the image with `microdnf upgrade` in it. I guess they do but without `microdnf upgrade` we take this question off the table.

So to sum it up - I think it does more harm than good. Since this is a new product release (and a tech preview), I think it's worth to verify my hypothesis and get rid of it.

File: productize.sh (hunk 2 above)
Comment (slaskawi) [reply productize.sh:6]: Fixed.

File: productize.sh (hunk 3 above)
Comment (slaskawi) [reply productize.sh:18]: Thanks for suggestion!

File: container.yaml (hunk 4 above)
Comment (slaskawi) [reply container.yaml]: Removed.

**Comment by slaskawi** [comment]
@hmlnarik Comments addressed.

File: Dockerfile (hunk 1 above)
Comment (slaskawi) [reply Dockerfile:13]: As per @hmlnarik request, re-introduced this piece.
